File: testing/reaper_project.py
#!/usr/bin/env python3
"""
REAPER project file (.rpp) generator for JSFX testing.
Creates minimal project files with test signals and JSFX effects loaded.
"""

from pathlib import Path
from typing import Dict, List, Optional


class ReaperProject:
    """Generate REAPER project files for automated testing."""
    
    def __init__(self, sample_rate=48000, bpm=120):
        """
        Initialize REAPER project generator.
        
        Args:
            sample_rate: Project sample rate
            bpm: Project tempo (BPM)
        """
        self.sample_rate = sample_rate
        self.bpm = bpm
        self.tracks = []
        
    def add_track_with_media(self, media_file, track_name="Test", jsfx_effects=None):
        """
        Add a track with media file and optional JSFX effects.
        
        Args:
            media_file: Path to audio file (WAV)
            track_name: Name of the track
            jsfx_effects: List of tuples (jsfx_path, slider_values_dict)
                         Example: [("LowPassFilter.jsfx", {"frequencySlider": 1000})]
        """
        track = {
            'name': track_name,
            'media_file': str(Path(media_file).absolute()),
            'effects': jsfx_effects or []
        }
        self.tracks.append(track)
        
    def _format_fx_chain(self, effects):
        """Format FX chain for .rpp file."""
        if not effects:
            return ""
        
        fx_lines = ["  <FXCHAIN"]
        
        for idx, (jsfx_path, sliders) in enumerate(effects):
            jsfx_name = Path(jsfx_path).stem
            fx_lines.append(f"    WNDRECT 0 0 0 0")
            fx_lines.append(f"    SHOW 0")
            fx_lines.append(f"    LASTSEL 0")
            fx_lines.append(f"    DOCKED 0")
            fx_lines.append(f"    <JS {jsfx_name} \"{jsfx_path}\"")
            
            # Add slider values
            slider_count = len(sliders) if sliders else 0
            slider_line = f"    {slider_count}"
            
            if sliders:
                # JSFX sliders are indexed starting from 1
                for slider_name, value in sliders.items():
                    slider_line += f" {value}"
            
            fx_lines.append(slider_line)
            fx_lines.append(f"    >")
        
        fx_lines.append("  >")
        return "\n".join(fx_lines)
    
    def generate_rpp(self, output_file, render_settings=None):
        """
        Generate .rpp project file.
        
        Args:
            output_file: Output .rpp filename
            render_settings: Optional dict with render settings
                           Keys: 'tail_ms' (render tail in ms, default 1000)
        """
        output_path = Path(output_file)
        output_path.parent.mkdir(parents=True, exist_ok=True)
        
        render_settings = render_settings or {}
        tail_ms = render_settings.get('tail_ms', 1000)
        
        # Build project content
        lines = [
            "<REAPER_PROJECT 0.1 \"7.0\" 1234567890",
            f"  SAMPLERATE {self.sample_rate} 0 0",
            f"  TEMPO {self.bpm} 4 4",
            "  MASTERAUTOMODE 0",
            "  MASTERVUMODE 0",
            "  MASTERTRACKHEIGHT 0",
            "  MASTERPEAKCOL 16576",
            "  RECORD_PATH \"\" \"\"",
            f"  RENDER_FILE \"\"",
            f"  RENDER_PATTERN \"\"",
            f"  RENDER_FMT 0 2 {self.sample_rate}",  # WAV, 16-bit, sample rate
            f"  RENDER_1X 0",
            f"  RENDER_RANGE 1 0 0 18 1000",  # Render project, time selection
            f"  RENDER_RESAMPLE 3 0 1",
            f"  RENDER_ADDTOPROJ 0",
            f"  RENDER_STEMS 0",
            f"  RENDER_DITHER 0",
            f"  TIMELOCKMODE 1",
            f"  RENDER_TAILFLAG 1",
            f"  RENDER_TAILMS {tail_ms}",
        ]
        
        # Add tracks
        for track_idx, track in enumerate(self.tracks):
            lines.append(f"  <TRACK {{{'{'}{track_idx:08X}-0000-0000-0000-000000000000'}}}")
            lines.append(f"    NAME \"{track['name']}\"")
            lines.append(f"    PEAKCOL 16576")
            lines.append(f"    VOLPAN 1 0 -1 -1 1")
            lines.append(f"    MUTESOLO 0 0 0")
            lines.append(f"    IPHASE 0")
            lines.append(f"    PLAYOFFS 0 1")
            lines.append(f"    ISBUS 0 0")
            lines.append(f"    BUSCOMP 0 0 0 0 0")
            lines.append(f"    SHOWINMIX 1 0.6667 0.5 1 0.5 0 0 0")
            lines.append(f"    SEL 0")
            lines.append(f"    REC 0 0 1 0 0 0 0 0")
            lines.append(f"    VU 2")
            lines.append(f"    TRACKHEIGHT 0 0 0 0 0 0")
            lines.append(f"    INQ 0 0 0 0.5 100 0 0 100")
            lines.append(f"    NCHAN 2")
            
            # Add FX chain
            fx_chain = self._format_fx_chain(track['effects'])
            if fx_chain:
                lines.append(fx_chain)
            
            # Add media item
            media_path = track['media_file']
            lines.append(f"    <ITEM")
            lines.append(f"      POSITION 0")
            lines.append(f"      LENGTH 10")  # Length will be auto-adjusted
            lines.append(f"      LOOP 0")
            lines.append(f"      ALLTAKES 0")
            lines.append(f"      FADEIN 1 0 0 1 0 0 0")
            lines.append(f"      FADEOUT 1 0 0 1 0 0 0")
            lines.append(f"      VOLPAN 1 0 1 -1")
            lines.append(f"      <SOURCE WAVE")
            lines.append(f"        FILE \"{media_path}\"")
            lines.append(f"      >")
            lines.append(f"    >")
            
            lines.append(f"  >")
        
        lines.append(">")
        
        # Write to file
        with open(output_path, 'w') as f:
            f.write('\n'.join(lines))
        
        return output_path


def create_test_project(jsfx_path, input_wav, output_rpp, slider_values=None, sample_rate=48000):
    """
    Quick helper to create a test project.
    
    Args:
        jsfx_path: Path to JSFX effect file
        input_wav: Path to input WAV file
        output_rpp: Path for output .rpp file
        slider_values: Dict of slider values (e.g., {"frequencySlider": 1000})
        sample_rate: Project sample rate
    
    Returns:
        Path to created .rpp file
    """
    project = ReaperProject(sample_rate=sample_rate)
    project.add_track_with_media(
        media_file=input_wav,
        track_name="Test Signal",
        jsfx_effects=[(jsfx_path, slider_values or {})]
    )
    return project.generate_rpp(output_rpp)


if __name__ == "__main__":
    # Example usage
    create_test_project(
        jsfx_path="LowPassFilter.jsfx",
        input_wav="test_signals/sine_1000hz.wav",
        output_rpp="test_projects/test_lowpass.rpp",
        slider_values={"frequencySlider": 500}
    )
    print("Created test project: test_projects/test_lowpass.rpp")

File: testing/reaper_project.lua
--[[
  REAPER project file (.rpp) generator for JSFX testing.
  Creates minimal project files with test signals and JSFX effects loaded.
  
  Lua port of reaper_project.py
]]

local ReaperProject = {}
ReaperProject.__index = ReaperProject

--- Create a new REAPER project generator
-- @param sample_rate Project sample rate (default 48000)
-- @param bpm Project tempo in BPM (default 120)
-- @return ReaperProject instance
function ReaperProject.new(sample_rate, bpm)
    local self = setmetatable({}, ReaperProject)
    self.sample_rate = sample_rate or 48000
    self.bpm = bpm or 120
    self.tracks = {}
    return self
end

--- Add a track with media file and optional JSFX effects
-- @param media_file Path to audio file (WAV)
-- @param track_name Name of the track (default "Test")
-- @param jsfx_effects Table of effects: {{path, sliders}, {path, sliders}, ...}
--                     sliders is a table of {name = value} pairs
function ReaperProject:add_track_with_media(media_file, track_name, jsfx_effects)
    local track = {
        name = track_name or "Test",
        media_file = media_file,
        effects = jsfx_effects or {}
    }
    table.insert(self.tracks, track)
end

--- Format FX chain for .rpp file
-- @param effects Table of effects
-- @return String containing FX chain section
local function format_fx_chain(effects)
    if not effects or #effects == 0 then
        return ""
    end
    
    local lines = {"  <FXCHAIN"}
    
    for idx, effect in ipairs(effects) do
        local jsfx_path = effect[1]
        local sliders = effect[2] or {}
        
        -- Extract just the filename from path
        local jsfx_name = jsfx_path:match("([^/\\]+)$") or jsfx_path
        jsfx_name = jsfx_name:gsub("%.jsfx$", "")
        
        table.insert(lines, "    WNDRECT 0 0 0 0")
        table.insert(lines, "    SHOW 0")
        table.insert(lines, "    LASTSEL 0")
        table.insert(lines, "    DOCKED 0")
        table.insert(lines, string.format('    <JS %s "%s"', jsfx_name, jsfx_path))
        
        -- Build slider line
        local slider_values = {}
        for name, value in pairs(sliders) do
            table.insert(slider_values, tostring(value))
        end
        
        local slider_count = #slider_values
        local slider_line = "    " .. slider_count
        if slider_count > 0 then
            slider_line = slider_line .. " " .. table.concat(slider_values, " ")
        end
        
        table.insert(lines, slider_line)
        table.insert(lines, "    >")
    end
    
    table.insert(lines, "  >")
    return table.concat(lines, "\n")
end

--- Ensure directory exists for output file
-- @param filepath Path to file
local function ensure_dir(filepath)
    local dir = filepath:match("(.*/)")
    if dir then
        os.execute('mkdir -p "' .. dir .. '"')
    end
end

--- Get absolute path
-- @param path Relative or absolute path
-- @return Absolute path
local function get_absolute_path(path)
    if path:sub(1, 1) == "/" then
        return path
    end
    -- Get current working directory
    local handle = io.popen("pwd")
    if handle then
        local cwd = handle:read("*a"):gsub("%s+$", "")
        handle:close()
        return cwd .. "/" .. path
    end
    return path
end

--- Generate .rpp project file
-- @param output_file Output .rpp filename
-- @param render_settings Optional table with render settings (tail_ms)
-- @return Path to created file
function ReaperProject:generate_rpp(output_file, render_settings)
    ensure_dir(output_file)
    
    render_settings = render_settings or {}
    local tail_ms = render_settings.tail_ms or 1000
    
    local lines = {
        '<REAPER_PROJECT 0.1 "7.0" 1234567890',
        string.format("  SAMPLERATE %d 0 0", self.sample_rate),
        string.format("  TEMPO %d 4 4", self.bpm),
        "  MASTERAUTOMODE 0",
        "  MASTERVUMODE 0",
        "  MASTERTRACKHEIGHT 0",
        "  MASTERPEAKCOL 16576",
        '  RECORD_PATH "" ""',
        '  RENDER_FILE ""',
        '  RENDER_PATTERN ""',
        string.format("  RENDER_FMT 0 2 %d", self.sample_rate),
        "  RENDER_1X 0",
        "  RENDER_RANGE 1 0 0 18 1000",
        "  RENDER_RESAMPLE 3 0 1",
        "  RENDER_ADDTOPROJ 0",
        "  RENDER_STEMS 0",
        "  RENDER_DITHER 0",
        "  TIMELOCKMODE 1",
        "  RENDER_TAILFLAG 1",
        string.format("  RENDER_TAILMS %d", tail_ms),
    }
    
    -- Add tracks
    for track_idx, track in ipairs(self.tracks) do
        local guid = string.format("{%08X-0000-0000-0000-000000000000}", track_idx - 1)
        table.insert(lines, string.format("  <TRACK %s", guid))
        table.insert(lines, string.format('    NAME "%s"', track.name))
        table.insert(lines, "    PEAKCOL 16576")
        table.insert(lines, "    VOLPAN 1 0 -1 -1 1")
        table.insert(lines, "    MUTESOLO 0 0 0")
        table.insert(lines, "    IPHASE 0")
        table.insert(lines, "    PLAYOFFS 0 1")
        table.insert(lines, "    ISBUS 0 0")
        table.insert(lines, "    BUSCOMP 0 0 0 0 0")
        table.insert(lines, "    SHOWINMIX 1 0.6667 0.5 1 0.5 0 0 0")
        table.insert(lines, "    SEL 0")
        table.insert(lines, "    REC 0 0 1 0 0 0 0 0")
        table.insert(lines, "    VU 2")
        table.insert(lines, "    TRACKHEIGHT 0 0 0 0 0 0")
        table.insert(lines, "    INQ 0 0 0 0.5 100 0 0 100")
        table.insert(lines, "    NCHAN 2")
        
        -- Add FX chain
        local fx_chain = format_fx_chain(track.effects)
        if fx_chain ~= "" then
            table.insert(lines, fx_chain)
        end
        
        -- Add media item
        local media_path = get_absolute_path(track.media_file)
        table.insert(lines, "    <ITEM")
        table.insert(lines, "      POSITION 0")
        table.insert(lines, "      LENGTH 10")
        table.insert(lines, "      LOOP 0")
        table.insert(lines, "      ALLTAKES 0")
        table.insert(lines, "      FADEIN 1 0 0 1 0 0 0")
        table.insert(lines, "      FADEOUT 1 0 0 1 0 0 0")
        table.insert(lines, "      VOLPAN 1 0 1 -1")
        table.insert(lines, "      <SOURCE WAVE")
        table.insert(lines, string.format('        FILE "%s"', media_path))
        table.insert(lines, "      >")
        table.insert(lines, "    >")
        
        table.insert(lines, "  >")
    end
    
    table.insert(lines, ">")
    
    -- Write to file
    local file = io.open(output_file, "w")
    if not file then
        error("Could not open file for writing: " .. output_file)
    end
    file:write(table.concat(lines, "\n"))
    file:close()
    
    return output_file
end

--- Quick helper to create a test project
-- @param jsfx_path Path to JSFX effect file
-- @param input_wav Path to input WAV file
-- @param output_rpp Path for output .rpp file
-- @param slider_values Table of slider values
-- @param sample_rate Project sample rate
-- @return Path to created .rpp file
local function create_test_project(jsfx_path, input_wav, output_rpp, slider_values, sample_rate)
    local project = ReaperProject.new(sample_rate or 48000)
    project:add_track_with_media(
        input_wav,
        "Test Signal",
        {{jsfx_path, slider_values or {}}}
    )
    return project:generate_rpp(output_rpp)
end

-- Export module
return {
    ReaperProject = ReaperProject,
    create_test_project = create_test_project
}
